src/vdom.rs

```diff
@@ -72,45 +103,50 @@ pub(crate) fn build(host: &Host, input: VDom<()>) -> VDomNode<Vec<VDom<Node>>, N
                 })
                 .collect();
             // Return the machine
-            VDomNode::Elem {
-                name,
-                attrs: attrs_new,
-                children,
-                state: node,
-            }
+            VDomNode::new_elem(elem.name, attrs_new, children, node)
         }
     }
 }
 
 impl VDomNode<Vec<VDom<Node>>, Node, Attrs> {
+    pub fn new_text(text: String, state: Node) -> Self {
+        VDomNode::Text(Text::new(text, state))
+    }
+
+    pub fn new_elem(name: String, attrs: Attrs, children: Vec<VDom<Node>>, state: Node) -> Self {
+        VDomNode::Elem(Elem::new(name, attrs, children, state))
+    }
     pub(crate) fn node(&self) -> &Node {
         match self {
-            VDomNode::Text { state, .. } => state,
-            VDomNode::Elem { state, .. } => state,
+            VDomNode::Text(text) => &text.state,
+            VDomNode::Elem(elem) => &elem.state,
         }
     }
 }
 
 impl VDom<Node> {
+    pub fn new(vdom: VDomNode<Vec<VDom<Node>>, Node, Attrs>) -> Self {
+        VDom { vdom }
+    }
+
     pub(crate) fn install(&self, host: &Host) {
         host.install(self.vdom.node());
     }
 
     pub(crate) fn halt(&mut self) -> Option<Node> {
-        match &mut self.vdom {
```

Comment:
If we don't use mutability inside halt, then do we need the `self` parameter to be mutable?

Reply:
there is mutation in the `VDomNode::Elem(...)` arm like the following
```rs
                elem.children.iter_mut().for_each(|x| {
                    x.halt();
                });
```

Reply:
Ah sure, I didn't realise that mutable manipulation was still possible without matching on `mut self.vdom`